src/data/move.ts

```diff
@@ -2875,6 +2875,164 @@ export class StatStageChangeAttr extends MoveEffectAttr {
   }
 }
 
+/**
+ * Attribute used to determine the Biome/Terrain-based secondary effect of Secret Power
+ */
+export class SecretPowerAttr extends MoveEffectAttr {
+  constructor() {
+    super(false);
+  }
+
+  /**
+   * Used to determine if the move should apply a secondary effect based on Secret Power's 30% chance
+   * @param user
+   * @param target
+   * @param move
+   * @param args
+   * @returns `true` if the move should | `false` if the roll fails and the move should not
+   */
+  override canApply(user: Pokemon, target: Pokemon, move: Move, args?: any[]): boolean {
+    this.effectChanceOverride = move.chance;
+    const moveChance = this.getMoveChance(user, target, move, this.selfTarget);
+    if (moveChance < 0 || moveChance === 100 || user.randSeedInt(100) < moveChance) {
+      this.effectChanceOverride = 100; // effectChanceOverride used in the application of the actual secondary effect
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  /**
+   * Used to apply the secondary effect to the target Pokemon
+   * @param user
+   * @param target
+   * @param move
+   * @param args
+   * @returns `true` if a secondary effect is successfully applied | 'false' if not
+   */
+  override apply(user: Pokemon, target: Pokemon, move: Move, args?: any[] | undefined): boolean | Promise<boolean> {
+    if (!super.apply(user, target, move, args)) {
+      return false;
+    }
+    let secondaryEffect: MoveEffectAttr;
+    const terrain = user.scene.arena.getTerrainType();
+    if (terrain !== TerrainType.NONE) {
+      secondaryEffect = this.determineTerrainEffect(terrain);
+    } else {
+      const biome = user.scene.arena.biomeType;
+      secondaryEffect = this.determineBiomeEffect(biome);
+    }
+    return secondaryEffect!.apply(user, target, move, []);
+  }
+
+  /**
+   * Determines the secondary effect based on terrain
+   * Takes precedence over biome
+   * Electric Terrain | Paralysis
+   * Misty Terrain    | SpAtk -1
+   * Grassy Terrain   | Sleep
+   * Psychic Terrain  | Speed -1
+   * @param terrain {@linkcode TerrainType}the current terrain
+   * @returns the chosen secondary effect {@linkcode MoveEffectAttr}
+   */
+  private determineTerrainEffect(terrain: TerrainType): MoveEffectAttr {
+    let secondaryEffect: MoveEffectAttr;
+    switch (terrain) {
+    case TerrainType.ELECTRIC:
+      secondaryEffect = new StatusEffectAttr(StatusEffect.PARALYSIS, false);
+      break;
+    case TerrainType.MISTY:
+      secondaryEffect = new StatStageChangeAttr([ Stat.SPATK ], -1, false);
+      break;
+    case TerrainType.GRASSY:
+      secondaryEffect = new StatusEffectAttr(StatusEffect.SLEEP, false);
+      break;
+    case TerrainType.PSYCHIC:
+      secondaryEffect = new StatStageChangeAttr([ Stat.SPD ], -1, false);
+      break;
+    }
+    return secondaryEffect!;
```

Comment:
```suggestion
    return secondaryEffect;
```
The reason the `!` was needed is because there was no `default` case for the switch statement. Adding a default case (which should be there anyway, to prevent issues if GameFreak adds more terrain types in the future and this move is missed) should fix that.